aws-lc-rs/src/rsa/key.rs

```diff
@@ -3,43 +3,94 @@
-    mem::MaybeUninit,
-    ptr::null_mut,
-};
-
-#[cfg(feature = "fips")]
-use aws_lc::RSA_check_fips;
-use aws_lc::{
-    EVP_DigestSignInit, EVP_PKEY_assign_RSA, EVP_PKEY_new, RSA_get0_e, RSA_get0_n, RSA_get0_p,
-    RSA_get0_q, RSA_new, RSA_parse_private_key, RSA_parse_public_key, RSA_public_key_to_bytes,
-    RSA_set0_key, RSA_size, EVP_PKEY, EVP_PKEY_CTX, RSA,
-};
-
-use mirai_annotations::verify_unreachable;
-
-#[cfg(feature = "ring-io")]
-use untrusted::Input;
-
-use zeroize::Zeroize;
-
 use super::{
+    encoding,
     signature::{compute_rsa_signature, RsaEncoding, RsaPadding},
     RsaParameters,
 };
-
 #[cfg(feature = "ring-io")]
 use crate::io;
+#[cfg(feature = "ring-io")]
+use crate::ptr::ConstPointer;
 use crate::{
-    cbs, digest,
+    buffer::Buffer,
+    digest::{self},
+    encoding::{AsDer, Pkcs8V1Der, RsaPublicKeyX509Der},
     error::{KeyRejected, Unspecified},
+    fips::indicator_check,
     hex,
-    ptr::{ConstPointer, DetachableLcPtr, LcPtr},
+    ptr::{DetachableLcPtr, LcPtr, Pointer},
     rand,
     sealed::Sealed,
 };
+#[cfg(feature = "fips")]
+use aws_lc::RSA_check_fips;
+use aws_lc::{
+    EVP_DigestSignInit, EVP_PKEY_assign_RSA, EVP_PKEY_bits, EVP_PKEY_id, EVP_PKEY_new,
+    EVP_PKEY_size, RSA_generate_key_ex, RSA_generate_key_fips, RSA_get0_d, RSA_new, RSA_set0_key,
+    RSA_size, BIGNUM, EVP_PKEY, EVP_PKEY_CTX, EVP_PKEY_RSA, RSA_F4,
+};
+#[cfg(feature = "ring-io")]
+use aws_lc::{RSA_get0_e, RSA_get0_n};
+use mirai_annotations::verify_unreachable;
+use std::{
+    fmt::{self, Debug, Formatter},
+    ptr::null_mut,
+};
+#[cfg(feature = "ring-io")]
+use untrusted::Input;
+use zeroize::Zeroize;
+
+macro_rules! rsa_key_size_enum {
+    ($name:ident) => {
+        /// RSA key-size constants.
+        #[allow(clippy::len_without_is_empty)]
+        #[non_exhaustive]
+        #[derive(Clone, Copy, Debug, PartialEq, Eq)]
+        pub enum $name {
+            /// 2048-bit key
+            Rsa2048,
+
+            /// 3072-bit key
+            Rsa3072,
+
+            /// 4096-bit key
+            Rsa4096,
+
+            /// 8192-bit key
+            Rsa8192,
+        }
+
+        impl $name {
+            /// Returns the size of the key in bytes.
+            #[inline]
+            #[must_use]
+            pub fn len(self) -> usize {
+                match self {
+                    Self::Rsa2048 => 256,
+                    Self::Rsa3072 => 384,
+                    Self::Rsa4096 => 512,
+                    Self::Rsa8192 => 1024,
+                }
+            }
+
+            /// Returns the key size in bits.
+            #[inline]
+            fn bit_len(self) -> i32 {
+                match self {
+                    Self::Rsa2048 => 2048,
+                    Self::Rsa3072 => 3072,
+                    Self::Rsa4096 => 4096,
+                    Self::Rsa8192 => 8192,
+                }
+            }
+        }
+    };
+}
+
+pub(super) use rsa_key_size_enum;
+
+rsa_key_size_enum!(SignatureKeySize);
```

Comment:
* Why do we need to distinguish `EncryptionKeySize` from `SignatureKeySize`?
* Should this instead be in "signature.rs"?  I think either both `EncryptionKeySize` and `SignatureKeySize` should be here, or neither.

Reply:
Originally I had it as a single KeySize type shared between the two. I wasn't sure if there would be a case in the future where we would want to add size asymmetrically to one and not the other. Kind of a moot points given that nothing stops you from using one RSA key bytes with the other. I can recombine them into a single type.